### ISSUE TITLE
pagerRef.current.setPage is now working in useEffect

### ISSUE THREAD

**Issue by ryskin** [computerjazz/react-native-infinite-pager#22]
pagerRef.current.setPage is now working in useEffect, only when used with setTimeout. Any way to make the initial index state (rather than use difference)?

thanks

**Comment by michelecocuccio** [comment]
+1

**Comment by computerjazz** [comment]
can you give me some more context on the issue? and a snack repro if possible? thanks!

**Comment by computerjazz** [comment]
> Any way to make the initial index state (rather than use difference)

Is this a separate issue? You want to initialize the pager with index other than 0? You're correct that the initial index is always 0, but what you _map that 0 to_ is up to you. Pager index 0 could map to anything — including an item in the middle of some data array you have in state — part of the intentional design of this pager is that its agnostic of the data it renders — its only job is to page through indices, starting at 0 and going negative into infinity, or positive into infinity.

**Comment by michelecocuccio** [comment]
Hello @computerjazz 

first of all thank you for the library, is very very good for what I am trying to achieve.
I am basically using infinite page to create a Instagram Like stories component. Each story has an ID which I can use as Index for the infinite-pager.

![Simulator Screen Recording - iPhone 14 Pro - 2023-11-09 at 18 45 09](https://github.com/computerjazz/react-native-infinite-pager/assets/8071249/b22de3b2-1f81-4a52-a26e-5d1a18b2f0cd)


Everything works fine except that I don't have a way to start from that specific index. I cloned your code and I was able to make it work by adding a `startAtIndex` prop but the only way to make it work is by doing

```
useEffect(() => {
    if (startAtIndex) {
      const timeout = setTimeout(() => {
        setPage(startAtIndex, { animated: false });
      }, 300;

      return () => clearTimeout(timeout);
    }
  }, [startAtIndex]);
```

(which I think is what was mentioned in the issue above). As you might notice this is quite an ugly workaround. That 300 is clearly an imperative value that must be avoided and in fact it works on IOS in my case but not in Android (which has decreased performances compare to IOS). By incrementing the timeOut of course it works but it's ugly as you see the first page and then it jumps to the selected.

I know this maybe is not in your design pattern, but it will be fantastic if there was a way to add it or if you could tell me how to achieve this as it will make me able to finish this feature for my project.

Thanks

**Comment by computerjazz** [comment]
Thanks for the details! I agree it would be a minor change to be able to set the initial index and I'll consider adding it as a prop. Is there a reason why you can't do something like:

```tsx
const renderPage = useCallback(({ index }) => {
  const indexWithOffset = index + startAtIndex
  const item = myData[indexWithOffset]
  // render your item...
}, [startAtIndex])

return (
      <InfinitePager
        renderPage={renderPage}
      />)
```

**Comment by michelecocuccio** [comment]
> ```tsx
> const renderPage = useCallback(({ index }) => {
>   const indexWithOffset = index + startAtIndex
>   const item = myData[indexWithOffset]
>   // render your item...
> }, [startAtIndex])
> ```

Thanks for the quick reply @computerjazz. Is nice to see someone actually replying. I have used a lot of react native library and they all seems abandoned and no ones ever reply. Anyway, yes this could work but then, it messes up the array. Meaning I start from the 2 second page but I can't go back to the previous ones

![Simulator Screen Recording - iPhone 14 Pro - 2023-11-10 at 08 23 09](https://github.com/computerjazz/react-native-infinite-pager/assets/8071249/41d6fea7-349a-41cf-851a-76924707174d)
![Simulator Screen Recording - iPhone 14 Pro - 2023-11-10 at 08 14 48](https://github.com/computerjazz/react-native-infinite-pager/assets/8071249/1a69a6ab-cdbc-4515-8081-fd79fd7e7f44)

**Comment by computerjazz** [comment]
as for the original issue  with `setPage` on mount -- should be fixed in `0.3.10`: https://github.com/computerjazz/react-native-infinite-pager/releases/tag/v0.3.10
